File: src/commonTest/kotlin/com/scoreo/application/DeletePlayerUseCaseTest.kt
package com.scoreo.application

import com.scoreo.FakePlayerRepository
import com.scoreo.domain.model.Player
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class DeletePlayerUseCaseTest {

    @Test
    fun `delete marks player as inactive`() {
        val repo = FakePlayerRepository()
        repo.save(Player("p1", "Alice"))
        val useCase = DeletePlayerUseCase(repo)

        useCase("p1")

        val all = repo.getAll(includeInactive = true)
        assertEquals(1, all.size)
        assertFalse(all.first().active)
    }

    @Test
    fun `delete keeps name by default`() {
        val repo = FakePlayerRepository()
        repo.save(Player("p1", "Alice"))
        val useCase = DeletePlayerUseCase(repo)

        useCase("p1")

        val all = repo.getAll(includeInactive = true)
        assertEquals("Alice", all.first().name)
    }

    @Test
    fun `delete with anonymize blanks name`() {
        val repo = FakePlayerRepository()
        repo.save(Player("p1", "Alice"))
        val useCase = DeletePlayerUseCase(repo)

        useCase("p1", anonymize = true)

        val all = repo.getAll(includeInactive = true)
        assertEquals("", all.first().name)
    }

    @Test
    fun `active players excluded by default after delete`() {
        val repo = FakePlayerRepository()
        repo.save(Player("p1", "Alice"))
        repo.save(Player("p2", "Bob"))
        val useCase = DeletePlayerUseCase(repo)

        useCase("p1")

        val active = repo.getAll()
        assertEquals(1, active.size)
        assertEquals("Bob", active.first().name)
    }

    @Test
    fun `delete non-existent id does not throw`() {
        val repo = FakePlayerRepository()
        val useCase = DeletePlayerUseCase(repo)

        useCase("non_existent")
        // should not throw
    }

    @Test
    fun `delete is idempotent`() {
        val repo = FakePlayerRepository()
        repo.save(Player("p1", "Alice"))
        val useCase = DeletePlayerUseCase(repo)

        useCase("p1")
        useCase("p1") // second delete

        val all = repo.getAll(includeInactive = true)
        assertEquals(1, all.size)
        assertFalse(all.first().active)
    }
}

File: apps/scoreo/src/application/deletePlayerUseCase.test.ts
import { describe, expect, it } from 'vitest'
import { InMemoryPlayerRepository } from '../infrastructure/testing/inMemoryPlayerRepository'
import { DeletePlayerUseCase } from './deletePlayerUseCase'

describe('DeletePlayerUseCase', () => {
  it('delete marks player as inactive', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: true })
    const useCase = new DeletePlayerUseCase(repo)

    useCase.invoke('p1')

    const all = repo.getAll(true)
    expect(all).toHaveLength(1)
    expect(all[0].active).toBe(false)
  })

  it('delete keeps name by default', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: true })
    const useCase = new DeletePlayerUseCase(repo)

    useCase.invoke('p1')

    expect(repo.getAll(true)[0].name).toBe('Alice')
  })

  it('delete with anonymize blanks name', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: true })
    const useCase = new DeletePlayerUseCase(repo)

    useCase.invoke('p1', true)

    expect(repo.getAll(true)[0].name).toBe('')
  })

  it('active players excluded by default after delete', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: true })
    repo.save({ id: 'p2', name: 'Bob', active: true })
    const useCase = new DeletePlayerUseCase(repo)

    useCase.invoke('p1')

    const active = repo.getAll()
    expect(active).toHaveLength(1)
    expect(active[0].name).toBe('Bob')
  })

  it('delete non-existent id does not throw', () => {
    const repo = new InMemoryPlayerRepository()
    const useCase = new DeletePlayerUseCase(repo)

    expect(() => useCase.invoke('non_existent')).not.toThrow()
  })

  it('delete is idempotent', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: true })
    const useCase = new DeletePlayerUseCase(repo)

    useCase.invoke('p1')
    useCase.invoke('p1')

    const all = repo.getAll(true)
    expect(all).toHaveLength(1)
    expect(all[0].active).toBe(false)
  })
})
